internal/backend/runtime/omni/controllers/omni/machine_extensions.go

```diff
@@ -0,0 +1,241 @@
+// Copyright (c) 2024 Sidero Labs, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+
+package omni
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cosi-project/runtime/pkg/controller"
+	"github.com/cosi-project/runtime/pkg/controller/generic"
+	"github.com/cosi-project/runtime/pkg/resource"
+	"github.com/cosi-project/runtime/pkg/safe"
+	"github.com/cosi-project/runtime/pkg/state"
+	"github.com/siderolabs/gen/optional"
+	"go.uber.org/zap"
+
+	"github.com/siderolabs/omni/client/pkg/omni/resources"
+	"github.com/siderolabs/omni/client/pkg/omni/resources/omni"
+	"github.com/siderolabs/omni/internal/backend/runtime/omni/controllers/helpers"
+)
+
+// MachineExtensionsControllerName is the name of the MachineExtensionsController.
+const MachineExtensionsControllerName = "MachineExtensionsController"
+
+// MachineExtensionsController splits a single extensions configuration resource defined for cluster/machine set
+// to the MachineExtensions resource for each machine.
+type MachineExtensionsController struct {
+	generic.NamedController
+}
+
+// NewMachineExtensionsController initializes MachineExtensionsController.
+func NewMachineExtensionsController() *MachineExtensionsController {
+	return &MachineExtensionsController{
+		NamedController: generic.NamedController{
+			ControllerName: MachineExtensionsControllerName,
+		},
+	}
+}
+
+// Settings implements controller.QController interface.
+func (ctrl *MachineExtensionsController) Settings() controller.QSettings {
+	return controller.QSettings{
+		Inputs: []controller.Input{
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ExtensionsConfigurationType,
+				Kind:      controller.InputQPrimary,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.ClusterMachineType,
+				Kind:      controller.InputQMapped,
+			},
+			{
+				Namespace: resources.DefaultNamespace,
+				Type:      omni.MachineExtensionsType,
```

Comment:
should it take it's output also as input (even as `DestroyReady`)?

Reply:
It needs to catch the moment when finalizers are moved from all produced `MachineExtensions`, so yes.
It doesn't work without this input.